### PR TITLE
Include pyenv and Python 3.9 to 3.12 in the kitchen sink image

### PR DESCRIPTION
**Pull request by julienp**
Use pyenv to install Python 3.9 to 3.12 in the image. The default version remains 3.9.

This is enough to make `.python-version` files work when they match the preinstalled versions, however it will fail to run python if the user specifies a patch version that is not installed, for example `3.9.0` (note that leaving out the patch version works, for example `3.9`). We need to include [support for installing these versions on demand](https://github.com/pulumi/pulumi/pull/16855) in `pulumi install`.

Ref https://github.com/pulumi/pulumi-docker-containers/issues/222

Fixes https://github.com/pulumi/pulumi-docker-containers/pull/232
Fixes https://github.com/pulumi/pulumi-docker-containers/issues/222

### CODE REVIEW

File: tests/testdata/python-3.10.2/.python-version [hunk 1]
```diff
@@ -0,0 +1 @@
+3.10.2
```
Comment (julienp) [inline tests/testdata/python-3.10.2/.python-version:1]: This test case will fail until we have a pulumi release with https://github.com/pulumi/pulumi/pull/16855

File: docker/pulumi/Dockerfile [hunk 2]
```diff
@@ -94,6 +105,23 @@ RUN curl -L https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-
...
+ENV PYENV_ROOT /pyenv
+ENV PATH="${PYENV_ROOT}/shims:${PYENV_ROOT}/bin:${PATH}"
+RUN pyenv install 3.12
+RUN pyenv install 3.11
+RUN pyenv install 3.10
+RUN pyenv install 3.9
+RUN pyenv global 3.9 # Default version
+# Poetry
+RUN curl -sSL https://install.python-poetry.org | python3 -
+ENV PATH="/root/.local/bin:${PATH}"
```
Comment (julienp) [inline docker/pulumi/Dockerfile:120]: Update this to /usr/local/share/poetry, see https://github.com/pulumi/pulumi-docker-containers/pull/240/files#diff-56dcbb4f608b8a8d5169091d1a21b391651622f6404e5ef885198868f23a2984R30

File: docker/pulumi/Dockerfile [hunk 3]
```diff
@@ -94,6 +105,23 @@ RUN curl -L https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-
   helm repo add stable https://charts.helm.sh/stable && \
   helm repo update
 
+# Python
+# Install Pyenv and preinstall supported Python versions
+RUN git clone --depth=1 https://github.com/pyenv/pyenv.git /pyenv
+ENV PYENV_ROOT /pyenv
```
Comment (julienp) [inline docker/pulumi/Dockerfile:111]: Update this to `/opt/pyenv`

File: docker/pulumi/Dockerfile [hunk 4]
```diff
@@ -103,13 +114,32 @@ RUN curl -L https://raw.githubusercontent.com/helm/helm/master/scripts/get-helm-
...
 # Passing --build-arg PULUMI_VERSION=vX.Y.Z will use that version
 # of the SDK. Otherwise, we use whatever get.pulumi.com thinks is
 # the latest
 ARG PULUMI_VERSION
 
 # Install the Pulumi SDK, including the CLI and language runtimes.
-RUN curl -fsSL https://get.pulumi.com/ | bash -s -- --version $PULUMI_VERSION && \
+# TODO: Should use $PULUMI_VERSION, not dev
+# RUN curl -fsSL https://get.pulumi.com/ | bash -s -- --version $PULUMI_VERSION && \
+RUN curl -fsSL https://get.pulumi.com | sh -s -- --version dev && \
```
Comment (julienp) [inline docker/pulumi/Dockerfile:142]: This needs to be updated once 3.130.0 is released, before we merge this.

**Review by tgummerer — APPROVED** [review]

**Pull request merged**